trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/GolemChcScript.java

```diff
@@ -0,0 +1,186 @@
+import de.uni_freiburg.informatik.ultimate.core.model.services.IUltimateServiceProvider;
+import de.uni_freiburg.informatik.ultimate.lib.smtlibutils.ManagedScript;
+import de.uni_freiburg.informatik.ultimate.logic.LoggingScript;
+import de.uni_freiburg.informatik.ultimate.logic.Logics;
+import de.uni_freiburg.informatik.ultimate.logic.Model;
+import de.uni_freiburg.informatik.ultimate.logic.Script;
+import de.uni_freiburg.informatik.ultimate.logic.Script.LBool;
+import de.uni_freiburg.informatik.ultimate.smtsolver.external.Executor;
+
+public class GolemChcScript implements IChcScript {
+	private static final boolean ADD_CLAUSE_NAMES = false;
+	private static final boolean ADD_COMMENTS = false;
+	private static final boolean DECLARE_FUNCTIONS = true;
+
+	private final IUltimateServiceProvider mServices;
+	private final ILogger mLogger;
+	private final ManagedScript mMgdScript;
+	private final long mDefaultQueryTimeout;
+
+	private boolean mProduceModels = false;
+
+	private LBool mLastResult;
+	private Model mLastModel = null;
+
+	public GolemChcScript(final IUltimateServiceProvider services, final ManagedScript mgdScript) {
+		this(services, mgdScript, -1L);
+	}
+
+	public GolemChcScript(final IUltimateServiceProvider services, final ManagedScript mgdScript,
+			final long defaultTimeout) {
+		mServices = services;
+		mLogger = services.getLoggingService().getLogger(getClass());
+		mMgdScript = mgdScript;
+		mDefaultQueryTimeout = defaultTimeout;
+	}
+
+	@Override
+	public Script getScript() {
+		return mMgdScript.getScript();
+	}
+
+	@Override
+	public LBool solve(final HcSymbolTable symbolTable, final List<HornClause> system) {
+		return solve(symbolTable, system, -1L);
+	}
+
+	@Override
+	public LBool solve(final HcSymbolTable symbolTable, final List<HornClause> system, final long timeout) {
+		// generate file with Horn clauses
+		final File tmpFile;
+		try {
+			tmpFile = File.createTempFile("golem_", ".smt2");
+			mLogger.info("Writing script to file " + tmpFile.getAbsolutePath());
+
+			final var dumperScript = new LoggingScript(tmpFile.getAbsolutePath(), false);
+			dumperScript.setLogic(Logics.HORN);
+
+			new ChcAsserter(mMgdScript, dumperScript, ADD_CLAUSE_NAMES, ADD_COMMENTS, DECLARE_FUNCTIONS)
+					.assertClauses(symbolTable, system);
+			dumperScript.checkSat();
+
+			dumperScript.exit();
+		} catch (final IOException e) {
+			throw new IllegalStateException(e);
+		}
+
+		try {
+			// run golem on file
+			final var executor = new Executor(getCommand(tmpFile), mMgdScript.getScript(), mLogger, mServices, "golem",
+					null, null, null, determineTimeout(timeout));
+
+			mLastResult = executor.parseCheckSatResult();
+			switch (mLastResult) {
+			case SAT:
+				mLastModel = mProduceModels ? executor.parseGetModelResult() : null;
+				break;
+			default:
+				mLastModel = null;
+				break;
+			}
+
+			return mLastResult;
+
+		} catch (final IOException e) {
+			throw new IllegalStateException(e);
```

Comment:
why IllegalStateException? This has nothing to do with state, right? Just wrap it in a RuntimeException.